### PR TITLE
package_managers: rpm: Support prefetching modules metadata

### PR DESCRIPTION
**Pull request by hlin**
In order to prefetch modules metadata, https://github.com/konflux-ci/rpm-lockfile-prototype/pull/33 generates the lock file with a new metadata field

```
---
lockfileVersion: 1
lockfileVendor: redhat
arches:
- arch: x86_64
  packages: [...]
  source: [...]
  # ↑ already exists, ↓ is new
  metadata:
  - url: <URL>                  # needed to know what to download
    repoid: <repoid>            # need to know where to put the file
    size: <size>                # integrity checking
    checksum: sha256:<checksum> # integrity checking
    mdtype: modules             # hint for calling modifyrepo
```

JIRA: RHELCMP-14016

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

### CODE REVIEW

**Comment by lubomir** [comment]
> With this patch, createrepo can recognize modules.yaml.gz and mdtype field (and modifyrepo) is not used, @lubomir can you take a look?

That's a good point. Createrepo will match on the file name. Realistically the URL will point to some existing repodata, where the filename will match what createrepo expects. The mdtype in the lockfile is indeed not necessary, so cachi2 could drop it.

File: cachi2/core/package_managers/rpm/main.py [hunk 1]
```diff
@@ -195,6 +195,17 @@ def _download(lockfile: RedhatRpmsLock, output_dir: Path) -> dict[Path, Any]:
             }
             Path.mkdir(dest.parent, parents=True, exist_ok=True)
 
+        for md in arch.metadata:
+            filename = Path(md.url).name.split("-", 2)[1]
```
Comment (lubomir) [inline cachi2/core/package_managers/rpm/main.py:199]: Is this required? Processing the filename would require the URL to point to a file that has the `<checksum>-modules.yaml.gz` format, which is not required. But createrepo might be fine even with the checksum being included.

Comment (hlin) [reply cachi2/core/package_managers/rpm/main.py:199]: Checked it again and it doesn't needed to remove the checksum.  While I can't remember what happened with previous tests.

File: cachi2/core/package_managers/rpm/main.py [hunk 2]
```diff
@@ -241,6 +251,8 @@ def _generate_sbom_components(
     """Fill the component list with the package records."""
     components: list[Component] = []
     for file_path, file_metadata in files_metadata.items():
+        if not str(file_path).endswith(".rpm"):
```
Comment (a-ovchinnikov) [inline cachi2/core/package_managers/rpm/main.py:254]: `file_path` should be a `Path`, you could get rid of a conversion here by checking its suffix directly:
```suggestion
if file_path.suffix != ".rpm":
```

This also is starting to get a bit complicated. I suggest shifting this to a helper called `is_rpm_metadata()` or something similar, otherwise it could be a bit unclear why non-rpm files are expected.

File: cachi2/core/package_managers/rpm/main.py [hunk 3]
```diff
@@ -195,6 +195,16 @@ def _download(lockfile: RedhatRpmsLock, output_dir: Path) -> dict[Path, Any]:
             }
             Path.mkdir(dest.parent, parents=True, exist_ok=True)
 
+        for md in arch.metadata:
```
Comment (a-ovchinnikov) [inline cachi2/core/package_managers/rpm/main.py:198]: Could you please also update the docstring to include an explanation of this new change in behavior?

**Review by a-ovchinnikov — COMMENTED** [review]
Could you please update the commit message to include some reason for this change? The current one just states that there is a extra field, a sentence or two about needing to fetch it would be helpful.

File: cachi2/core/package_managers/rpm/main.py [hunk 4]
```diff
@@ -359,11 +369,20 @@ def raise_exception(message: str) -> None:
                 raise_exception(f"Unmatched checksum of '{file_path}' != '{digest}'")
 
 
+def _is_rpm_file(file_path: Path):
+    """Check if it's a rpm file."""
+    if file_path.suffix == ".rpm":
```
Comment (a-ovchinnikov) [inline cachi2/core/package_managers/rpm/main.py:374]: This is a good helper method! It could be made even simpler:
```suggestion
return file_path.suffix == ".rpm"
```

File: cachi2/core/package_managers/rpm/main.py [hunk 5]
```diff
@@ -319,6 +319,16 @@ def _download(lockfile: RedhatRpmsLock, output_dir: Path) -> dict[Path, Any]:
             }
             Path.mkdir(dest.parent, parents=True, exist_ok=True)
 
+        for md in arch.metadata:
```
Comment (a-ovchinnikov) [inline cachi2/core/package_managers/rpm/main.py:322]: This function was refactored, could you please update your change to match the new code?

**Comment by eskultety** [comment]
I confirm that this approach which is consistent with the rest of the lockfile based logic and behaviour works. I tested this with the following minimalistic lockfile and Fedora 38 (39+ dropped modularity):
```
---
lockfileVersion: 1
lockfileVendor: redhat
arches:
- arch: x86_64
  packages:
  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Modular/x86_64/os/Packages/s/subversion-libs-1.14.2-5.module_f38+15272+edcc0b97.x86_64.rpm
    repoid: fedora-modular
    size: 1583542
    checksum: sha256:b52f95619aa4ee70190fce7837985504550938fe0c0ebfba0f2ccd94b9a0a897
    name: subversion-libs

  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Everything/x86_64/os/Packages/a/apr-1.7.2-2.fc38.x86_64.rpm
    repoid: fedora
    name: apr
    size: 130140
    checksum: sha256:ca4680178a8d7d5562dfbd361bc7caa13f8fc9a2756d337cf9ded77687e763ee

  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Everything/x86_64/os/Packages/a/apr-util-1.6.3-2.fc38.x86_64.rpm
    name: apr-util
    repoid: fedora
    size: 97926
    checksum: sha256:ecad94eb8493689ac063df1d4e141ea992996d1a1a73f7a6b77724d93b4063c0

  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Everything/x86_64/os/Packages/l/libserf-1.3.9-27.fc38.x86_64.rpm
    name: libserf
    repoid: updates
    size: 57965
    checksum: sha256:d04afa98de14d4a30ad8f4c335f0fa458ec89596a5bf6848718989ff03755363

  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Everything/x86_64/os/Packages/u/utf8proc-2.7.0-4.fc38.x86_64.rpm
    name: utf8proc
    repoid: fedora
    size: 82281
    checksum: sha256:5b4cdd58179acda1cbeedfbacba872d9e43f39c629486860420cf578b30c9a2c

  metadata:
  - url: https://archives.fedoraproject.org/pub/archive/fedora/linux/releases/38/Modular/x86_64/os/repodata/a5b951a984ab6b7edd2c08d68162014d9f6ccfdbc4c721d318edca0a21697cb1-modules.yaml.gz
    repoid: fedora-modular
    mdtype: modules
    size: 31546
    checksum: sha256:a5b951a984ab6b7edd2c08d68162014d9f6ccfdbc4c721d318edca0a21697cb1
```

Then, after having run `fetch --dev-package-mangers rpm` and `inject-files cachi2-output --for-output-dir /tmp/`
```
$ podman run -it --rm --network=none -v cachi2-output/deps/rpm/x86_64/repos.d/:/etc/yum.repos.d/:rw,Z -v cachi2-output/deps:/tmp/deps:rw,Z fedora:38 /bin/bash

# dnf module enable -y subversion:1.14
==========================================================================================================
 Package                  Architecture            Version                  Repository                Size
==========================================================================================================
Enabling module streams:
 subversion                                       1.14                                                   

Transaction Summary
==========================================================================================================
Complete!

# dnf install subversion-libs -y
Repository 'fedora-modular' is missing name in configuration, using id.
Repository 'fedora' is missing name in configuration, using id.
Repository 'updates' is missing name in configuration, using id.
Last metadata expiration check: 0:00:06 ago on Wed Aug 28 11:28:29 2024.
Dependencies resolved.
==========================================================================================================
 Package               Architecture Version                                    Repository            Size
==========================================================================================================
Installing:
 subversion-libs       x86_64       1.14.2-5.module_f38+15272+edcc0b97         fedora-modular       1.5 M
Installing dependencies:
 apr                   x86_64       1.7.2-2.fc38                               fedora               127 k
 apr-util              x86_64       1.6.3-2.fc38                               fedora                96 k
 libserf               x86_64       1.3.9-27.fc38                              updates               57 k
 utf8proc              x86_64       2.7.0-4.fc38                               fedora                80 k

Transaction Summary
==========================================================================================================
Install  5 Packages

Total size: 1.9 M
Installed size: 5.6 M
Is this ok [y/N]: y
Downloading Packages:
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                  1/1 
  Installing       : apr-1.7.2-2.fc38.x86_64                                                          1/5 
  Installing       : apr-util-1.6.3-2.fc38.x86_64                                                     2/5 
  Installing       : libserf-1.3.9-27.fc38.x86_64                                                     3/5 
  Installing       : utf8proc-2.7.0-4.fc38.x86_64                                                     4/5 
  Installing       : subversion-libs-1.14.2-5.module_f38+15272+edcc0b97.x86_64                        5/5 
  Running scriptlet: subversion-libs-1.14.2-5.module_f38+15272+edcc0b97.x86_64                        5/5 
  Verifying        : subversion-libs-1.14.2-5.module_f38+15272+edcc0b97.x86_64                        1/5 
  Verifying        : apr-1.7.2-2.fc38.x86_64                                                          2/5 
  Verifying        : apr-util-1.6.3-2.fc38.x86_64                                                     3/5 
  Verifying        : utf8proc-2.7.0-4.fc38.x86_64                                                     4/5 
  Verifying        : libserf-1.3.9-27.fc38.x86_64                                                     5/5 

Installed:
  apr-1.7.2-2.fc38.x86_64               apr-util-1.6.3-2.fc38.x86_64                                      
  libserf-1.3.9-27.fc38.x86_64          subversion-libs-1.14.2-5.module_f38+15272+edcc0b97.x86_64         
  utf8proc-2.7.0-4.fc38.x86_64         

Complete!
```

File: cachi2/core/package_managers/rpm/main.py [hunk 6]
```diff
@@ -286,9 +286,9 @@ def _resolve_rpm_project(source_dir: RootedPath, output_dir: RootedPath) -> list
 
 def _download(lockfile: RedhatRpmsLock, output_dir: Path) -> dict[Path, Any]:
     """
-    Download packages mentioned in the lockfile.
+    Download packages and metadata mentioned in the lockfile.
```
[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/rpm/main.py:289]: IIUC the type of metadata is denoted by the `mdtype` field but I don't see it used anywhere in the code

[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/rpm/main.py:289]: Oh, I just noticed the answer in https://github.com/containerbuildsystem/cachi2/pull/610#issuecomment-2309657451, so indeed it should be dropped.

[the remark this answers is not in the document]
Comment (lubomir) [reply cachi2/core/package_managers/rpm/main.py:289]: The approach would (almost) work for comps and updateinfo. Those are not automatically picked up by createrepo though. Getting them into the final repo requires an additional option or call to modifyrepo.

However, unlike the modular metadata, those two could be replaced with an alternative way to achieve whatever they do. So yup, let's get rid of mdtype and keep things simpler.

[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/rpm/main.py:289]: Sounds like it should be handled by DNF directly given the complexity.

File: cachi2/core/package_managers/rpm/main.py [hunk 7]
```diff
@@ -359,11 +360,18 @@ def raise_exception(message: str) -> None:
                 raise_exception(f"Unmatched checksum of '{file_path}' != '{digest}'")
 
 
+def _is_rpm_file(file_path: Path):
+    """Check if it's a rpm file."""
+    return file_path.suffix == ".rpm"
+
+
```
[the remark this answers is not in the document]
Comment (a-ovchinnikov) [reply cachi2/core/package_managers/rpm/main.py:367]: I will politely disagree. A simple boolean comparison tells a maintainer how one achieves some goal, a well-named function tells what the goal is. Immediately knowing a goal reduces mental load and makes code more straightforward. Here the goal is to check if some object refers to an rpm file and it happens that it boils down to a simple comparison. If at some point in time this becomes inadequate to the goal then it could be easily adjusted.

The overhead of a function call in this specific case (in most other cases too) is negligible, this is not a tight loop, the code above deals with downloading data which will always dominate by several orders of magnitude. Also this is Python after all, it is usually valued for readability, not for lack of overhead. I would still suggest keeping this function.

[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/rpm/main.py:367]: What you argue about makes sense...when there are multiple places where such a check could be employed, there's literally a single spot where this is used in a very distinct context, so I'm still not convinced.

[the remark this answers is not in the document]
Comment (eskultety) [reply cachi2/core/package_managers/rpm/main.py:367]: It's not a blocker for me though.

File: tests/unit/package_managers/test_rpm.py [hunk 8]
```diff
@@ -41,6 +41,11 @@
         checksum: sha256:94803b5e1ff601bf4009f223cb53037cdfa2fe559d90251bbe85a3a5bc6d2aab
         size: 14735448
         repoid: updates-source
+    metadata:
```
[the remark this answers is not in the document]
Comment (hlin) [reply tests/unit/package_managers/test_rpm.py:44]: I'm trying to update e2e test with ubi8 image.

[the remark this answers is not in the document]
Comment (hlin) [reply tests/unit/package_managers/test_rpm.py:44]: Added an e2e testcase for module https://github.com/hlin/cachi2-rpm/tree/e2e-module

[the remark this answers is not in the document]
Comment (eskultety) [reply tests/unit/package_managers/test_rpm.py:44]: Great, can you please try looking for a module with fewer dependencies to decrease the time needed for the test to run (i.e. decrease the download time), it's not a coincidence I used subversion libs in my Fedora example, because that one had only 4 deps and the packages are just a few KiB each.

[the remark this answers is not in the document]
Comment (eskultety) [reply tests/unit/package_managers/test_rpm.py:44]: You can then open a PR against the testing repo to merge the case addition first and then reference it from this PR.

[the remark this answers is not in the document]
Comment (hlin) [reply tests/unit/package_managers/test_rpm.py:44]: Updated https://github.com/hlin/cachi2-rpm/tree/e2e-module to install httpd-tools instead, which only needs to download around 400 KiB in total.
And can you help to create `e2e-module` branch https://github.com/cachito-testing/cachi2-rpm? and then I can create PR with it.

[the remark this answers is not in the document]
Comment (eskultety) [reply tests/unit/package_managers/test_rpm.py:44]: > And can you help to create e2e-module branch https://github.com/cachito-testing/cachi2-rpm? and then I can create PR with it.

Sure, I will create it if needed, but hear me out first.
We could further decrease the amount of time needed for the e2e by combining the modularity test with the existing e2e. Not only would it build a single container only, but it would also test the weird corner case scenario (covered by a conventional int. test case already) where you could have multiple "packages" inside a repo, IOW you'd have multiple lockfiles that we'd need to process, one lockfile would contain regular RPM deps and the other would be scoped for modularity.

I'd like to see it consolidated, but admittedly, that's an improvement that can be done as a follow-up, so if you don't feel up for the challenge, let me know and I create a new branch for you in the test repo.

[the remark this answers is not in the document]
Comment (hlin) [reply tests/unit/package_managers/test_rpm.py:44]: > Sure, I will create it if needed, but hear me out first. We could further decrease the amount of time needed for the e2e by combining the modularity test with the existing e2e. Not only would it build a single container only, but it would also test the weird corner case scenario (covered by a conventional int. test case already) where you could have multiple "packages" inside a repo, IOW you'd have multiple lockfiles that we'd need to process, one lockfile would contain regular RPM deps and the other would be scoped for modularity.

I don't understand it clearly that how to build single container with multiple lockfiles

> 
> I'd like to see it consolidated, but admittedly, that's an improvement that can be done as a follow-up, so if you don't feel up for the challenge, let me know and I'll create a new branch for you in the test repo.

Initially, I attempted to update the Containerfile and lockfile directly in the e2e branch. However, as you know, F39+ does not support modularity, which caused significant changes to test_data/rpm_e2e_test. I felt that these changes were not so ideal and then created a new test case with e2e-module branch.  I'd prefer to improve it as a follow-up.

[the remark this answers is not in the document]
Comment (eskultety) [reply tests/unit/package_managers/test_rpm.py:44]: Sorry for a long delay in response, I've created this branch for you: https://github.com/cachito-testing/cachi2-rpm/tree/e2e-modularity

[the remark this answers is not in the document]
Comment (hlin) [reply tests/unit/package_managers/test_rpm.py:44]: Updated integration test against https://github.com/cachito-testing/cachi2-rpm/pull/3

**Comment by hlin** [comment]
Can someone help to trigger the workflows?

**Comment by hlin** [comment]
Are there any more blockers to merge this PR?

File: cachi2/core/package_managers/rpm/main.py [hunk 9]
```diff
@@ -286,9 +286,9 @@ def _resolve_rpm_project(source_dir: RootedPath, output_dir: RootedPath) -> list
 
 def _download(lockfile: RedhatRpmsLock, output_dir: Path) -> dict[Path, Any]:
     """
-    Download packages mentioned in the lockfile.
```
Comment (eskultety) [inline cachi2/core/package_managers/rpm/main.py:289]: **Commit message:**

Please don't include the JIRA IDs in commit messages, they don't help anyone without access, so if those provide any further context it won't be available publicly.

File: cachi2/core/package_managers/rpm/redhat.py [hunk 10]
```diff
@@ -17,12 +17,19 @@ class LockfilePackage(BaseModel):
     size: Optional[int] = None
 
 
+class LockfileModuleMetadata(LockfilePackage):
+    """Module metadata item; represents module metadata file."""
+
+    repoid: str
+
```
Comment (eskultety) [inline cachi2/core/package_managers/rpm/redhat.py:24]: why aren't url, size, and checksum mentioned here? Especially `url` since it's a mandatory field without which we can't download anything. What about the rest, are those optional just like for a regular RPM?

Comment (hlin) [reply cachi2/core/package_managers/rpm/redhat.py:24]: Those fields are not mentioned because the only difference from the regular PRM is that a repoid is required.

Comment (eskultety) [reply cachi2/core/package_managers/rpm/redhat.py:24]: But if I omit a URL in the lockfile, how is cachi2 supposed to download the metadata? The URL field must be required as well just like for a package.

Comment (hlin) [reply cachi2/core/package_managers/rpm/redhat.py:24]: As it's already 'required' in the parent class, cachi2 will raise error.

Error: PackageManagerError: RPM lockfile 'rpms.lock.yaml' format is not valid: '('arches', 0, 'module_metadata', 0, 'url'): Field required'
  Check the correct format and whether any keys are missing in the lockfile.

Comment (eskultety) [reply cachi2/core/package_managers/rpm/redhat.py:24]: Darn, I overlooked the inherited model, all good.

**Review by eskultety — APPROVED** [review]

**Comment by hlin** [comment]
We have a target end date of Sep 11 for this feature. Can this PR be merged then?

**Comment by eskultety** [comment]
> We have a target end date of Sep 11 for this feature. Can this PR be merged then?

Release is planned around 17th-18th, we'll get it in by then.
@brunoapimentel or @taylormadore  can we get one more ACK here?

**Review by a-ovchinnikov — APPROVED** [review]

**Review by brunoapimentel — APPROVED** [review]
